s2spy/calendar_shifter.py

```diff
@@ -0,0 +1,188 @@
+    # pylint: disable=protected-access
+    if isinstance(interval.gap, str):
+        gap_time_dict = interval._parse_timestring(interval.gap)
+    else:
+        gap_time_dict = interval.gap
+    # pylint: disable=protected-access
+    shift_time_dict = interval._parse_timestring(shift)
+    # make the shift negative for the precursor to shift forward in time
+    if interval.role == "precursor":
+        shift_time_dict.update(
+            (key, value * -1) for key, value in shift_time_dict.items()
+        )
+    # combine gap and shift time dict
+    gap_shifted = {
+        k: gap_time_dict.get(k, 0) + shift_time_dict.get(k, 0)
+        for k in set(gap_time_dict) | set(shift_time_dict)
+    }
+
+    return gap_shifted
+
+
+def cal_shifter(
+    calendar: s2spy.time.Calendar, shift: Union[str, dict]
+) -> s2spy.time.Calendar:
+    """
+    Shift a Calendar instance by a given time offset. Instead of shifting the
+    anchor date, this function shifts two things in reference to the anchor date:
+    - the target period(s): as a gap between the anchor date and the start of the first
+        target period
+    - the precursor period(s): as a gap between the anchor date and the start of the
+        first precursor period
+    This way, the anchor year from the input calendar is maintained on the returned
+    calendar. This is important for train-test splitting at later stages.
+    - Args:
+        - calendar : a s2spy.time.Calendar instance
+        - shift: a pandas-like
+        frequency string (e.g. "10d", "2W", or "3M"), or a pandas.DateOffset
+        compatible dictionary such as {days=10}, {weeks=2}, or {months=1, weeks=2}
+    - Example:
+        Shift a calendar by a given dateoffset.
+        >>> import s2spy.time
+        >>> gap_shifted = gap_shift()
+        >>> cal = nwp_calendar_builder(start_date, '7d', '7d', precur_leadtime='14d', n_precurperiods=4)
+        >>> cal_shifted = shifter(cal, '7d')
+        Calendar(
+            anchor='07-01',
+            allow_overlap=False,
+            mapping=None,
+            intervals=[
+                Interval(role='target', length='7d', gap={'days': 7}),
+                Interval(role='precursor', length='7d', gap={'days': 7}),
+                Interval(role='precursor', length='7d', gap='0d'),
+                Interval(role='precursor', length='7d', gap='0d'),
+                Interval(role='precursor', length='7d', gap='0d'),
+            ]
+        )
+    """
+    # get 1st target interval and 1st precursor interval from calendar
+    target_interval = calendar.targets[0]
+    precursor_interval = calendar.precursors[0]
+
+    # shift gaps
+    target_gap_shifted = _gap_shift(target_interval, shift)
+    precursor_gap_shifted = _gap_shift(precursor_interval, shift)
+
+    # set the new gaps on the calendar
+    calendar_shifted = copy.deepcopy(calendar)
+    calendar_shifted.targets[0].gap = target_gap_shifted
+    calendar_shifted.precursors[0].gap = precursor_gap_shifted
```

Comment:
Let's get rid of the intermediate variables and comments on self-explanatory code.

```suggestion
    calendar_shifted = copy.deepcopy(calendar)
    calendar_shifted.targets[0].gap = _gap_shift(calendar.targets[0], shift)
    calendar_shifted.precursors[0].gap = _gap_shift(calendar.precursors[0], shift)
```